docs/modules/ROOT/pages/references/architecture/upgrade_controller.adoc

```diff
@@ -0,0 +1,323 @@
+    desiredUpdate: {} <2>
+    upstream: https://api.openshift.com/api/upgrades_info/v1/graph
+----
+<1> Template for the `ClusterVersion/version` object.
+<2> The `desiredUpdate` is ignored and set by the `UpgradeJob` controller.
+
+=== UpgradeConfig
+
+The `UpgradeConfig` CRD defines the upgrade schedule and the upgrade job template.
+The reconciliation loop of the controller creates `UpgradeJob` objects based on the `UpgradeConfig` object.
+
+[source,yaml]
+----
+apiVersion: managedupgrade.appuio.io/v1beta1
+kind: UpgradeConfig
+metadata:
+  name: cluster-upgrade
+spec:
+  schedule: <1>
+    cron: "0 22 * * 2"
+    isoWeek: "@odd"
+    location: "Europe/Zurich"
+    suspend: false <4>
+  pinVersionWindow: "4h" <2>
+  jobTemplate:
+    upgradeTimeout: "2h" <3>
+    preUpgradeHealthChecks: <4>
+      timeout: "30m"
+      checkCriticalAlerts: true
+      checkDegradedOperators: true
+      excludeAlerts:
+      - alertname: "KubePodCrashLooping"
+      excludeNamespaces:
+      - openshift-console
+      excludeOperators:
+      - openshift-monitoring
+      customQueries:
+      - query: "up{job=~"^argocd-.+$",namespace="syn"} != 1"
+    postUpgradeHealthChecks: <5>
+      timeout: "30m"
+      checkCriticalAlerts: true
+      checkDegradedOperators: true
+      excludeAlerts:
+      - alertname: "KubePodCrashLooping"
+      excludeNamespaces:
+      - openshift-console
+      excludeOperators:
+      - openshift-monitoring
+      customQueries:
+      - query: "up{job=~"^argocd-.+$",namespace="syn"} != 1"
+    webhooks: <6>
+      - url: "https://example.com/webhook"
+        annotations:
+          cluster_id: "bar"
+          tenant_id: "foo"
+  webhooks: <6>
+    - url: "https://example.com/webhook"
+      annotations:
+        cluster_id: "bar"
+        tenant_id: "foo"
+----
+<1> The upgrade schedule as defined in <<upgrade-schedule>>.
+<2> The time window before the maintenance window in which the upgrade version is pinned.
+UpgradeJobs are created at `schedule - pinVersionWindow`.
+<3> The timeout for the upgrade.
+The upgrade is marked as failed if it takes longer than this.
+<4> The health checks to perform before the upgrade as defined in <<upgrade-health-checks>>.
+<5> The health checks to perform after the upgrade as defined in <<upgrade-health-checks>>.
+<6> The webhook to send notifications to as defined in <<upgrade-webhook>>.
+Having multiple webhooks allows to send notifications to different systems.
+Both the `UpgradeConfig` and the `UpgradeJob` have a `webhooks` field since both might send notifications.
+
+=== UpgradeJob
+
+An `UpgradeJob` is created for each upgrade.
+It contains a snapshot of the most recent version in the `.status.availableUpdates` field, a snapshot of the config, and a timestamp when the upgrade should start.
+
+[source,yaml]
+----
+apiVersion: managedupgrade.appuio.io/v1beta1
+kind: UpgradeJob
+metadata:
+  name: cluster-upgrade-2021-01-01T22:00:00+01:00
```

Comment:
This isn't a valid object name. Would it make sense to generate names with a hash since the object contains the `upgradeAt` information anyway?